### PR TITLE
Add Storybook 8 tests

### PR DESCRIPTION
**Pull request by valentinpalkovic**
This PR continues and supersedes https://github.com/vitejs/vite-ecosystem-ci/pull/137.

Unfortunately, I couldn't push to @IanVS's branch, so I have opened a new PR.

Additionally to what Ian has already introduced to allow a repository subdirectory to be accepted, I have adjusted the `Task` interface a bit to pass in the `agent` so that the proper package manager's functions to install or execute tasks can be used inside the task hooks like `beforeTest` or `test`.

### CODE REVIEW

File: tests/storybook.ts [hunk 1]
```diff
@@ -1,12 +1,22 @@
...
+		branch: 'next',
+		build: ['build-storybook'],
+		beforeTest: async (agent) => {
+			$`${getCommand(agent, 'execute', ['http-server', 'storybook-static', '--port 6006', '--silent'])}`
+			await $`${getCommand(agent, 'execute', ['wait-on', 'http://localhost:6006'])}`
+		},
+		test: async (agent) => {
+			await $`${getCommand(agent, 'execute', ['@storybook/test-runner'])}`
+			await $`${getCommand(agent, 'execute', ['fkill-cli', ':6006'])}`
+		},
```
Comment (patak-dev) [inline tests/storybook.ts:20]: Thanks for working on this PR! It would be great to have Storybook in our CI again!

About these scripts, having the `agent` looks ok to me to handle edge cases but the details in `beforeTest` and `test` will end up creating maintenance issues because you'll need to update them here instead of being in sync with your repo. Would it be possible to have a `test:vite` script in your repo that does all this so you can maintain it and change it independently of vite-ecosystem-ci?

Comment (valentinpalkovic) [reply tests/storybook.ts:20]: The https://github.com/storybookjs/sandboxes repository is used to build Storybook for different frameworks and renderer combinations. It is updated once per day. We run a script like `npm create vite` and initialize Storybook on top of it. Our scripts are currently not very flexible enough to hook in after build and, for example, adjust the package.json of a particular sandbox and set everything up to provide a `test:vite` script command.

Second, our users also use these sandboxes to provide reproductions. So, creating a script like `test:vite` would expose it to our users as well, which we potentially don't want to do.

I understand your concerns! I am optimistic that how a working Storybook can be tested will not change frequently. Running Storybook on the port `6006` wasn't ever changed and is unlikely to be changed. The test-runner is also a package that we maintain, and the CLI should remain stable as it is.

If you still have concerns, I suggest adjusting our sandbox generation scripts to hook in after the build. But again, the sandboxes should mirror real-world scenarios, and there, the `@storybook/test-runner`, for example, isn't set up per default.

Comment (patak-dev) [reply tests/storybook.ts:20]: This sounds ok to me for now then. @dominikg, what do you think? If we see this is causing issues we can rethink how we are integrating with Storybook later

File: tests/storybook.ts [hunk 2]
```diff
@@ -1,12 +1,22 @@
...
-		repo: 'storybookjs/builder-vite',
-		branch: 'main',
-		build: 'prepublish',
-		test: ['build-examples', 'test-ci'],
+		repo: 'storybookjs/sandboxes',
+		dir: 'next',
+		testSubdirectory: 'react-vite/default-ts/after-storybook',
+		branch: 'next',
+		build: ['build-storybook'],
+		beforeTest: async (agent) => {
```
Comment (dominikg) [inline tests/storybook.ts:13]: do we need to pass through agent here? a project should know it's preferred agent so just using `pnpm exeute...` here should work?

File: tests/storybook.ts (hunk 1 above)
Comment (dominikg) [reply tests/storybook.ts:20]: i'd still prefer this to be encapsulated in a script controlled on the generated projects side. Long term we need to redesign ecosystem-ci to have less configuration about the projects under test in this repo.

Is there a way for the storybook testrunner to start/stop the webserver itself similar to playwright? https://playwright.dev/docs/test-webserver  ? That could reduce the calls here to a single one.

the hardcoded port 6006 (which i believe is the default port of storybook) could also be a problem when running ecosystem-ci locally. This is needed to be able to reproduce or bisect issues.

File: tests/storybook.ts [hunk 3]
```diff
@@ -1,12 +1,22 @@
...
 export async function test(options: RunOptions) {
 	await runInRepo({
 		...options,
-		repo: 'storybookjs/builder-vite',
-		branch: 'main',
-		build: 'prepublish',
-		test: ['build-examples', 'test-ci'],
+		repo: 'storybookjs/sandboxes',
+		dir: 'next',
+		testSubdirectory: 'react-vite/default-ts/after-storybook',
```
Comment (dominikg) [inline tests/storybook.ts:10]: is there a reason to pick this specific subdirectory, i suppose there are multiple setups with vite in storybooks tests. Is this the one with the biggest test coverage or quickest runtime? I'm totally fine with making a compromise here to reduce execution time. But if other subdirectories would increase coverage, is there a way to add them?

Regarding the testSubdirectory, could this be prevented as well with using `pnpm --dir <testSubDirectory>` calls in the script block directly? or in the `"scripts"` section of the storbookjs/sandboxes monorepo root?

**Comment by valentinpalkovic** [comment]
Superseded by https://github.com/vitejs/vite-ecosystem-ci/pull/311

File: tests/storybook.ts (hunk 3 above)
Comment (valentinpalkovic) [reply tests/storybook.ts:10]: We have different sandboxes testing renderer/builder and TypeScript/JavaScript combinations. These are all our sandboxes, which we have in place using the Vite builder:

- react-vite/default-js
- react-vite/default-ts
- react-vite/prerelease-ts
- solid-vite/default-js
- solid-vite/default-ts
- vue3-vite/default-js
- vue3-vite/default-ts
- html-vite/default-js
- html-vite/default-ts
- svelte-vite/default-js
- svelte-vite/default-ts
- svelte-kit/skeleton-js
- svelte-kit/skeleton-ts
- svelte-kit/prerelease-ts
- lit-vite/default-js
- lit-vite/default-ts
- preact-vite/default-js
- preact-vite/default-ts
